Add separate expirable LRU implementation, with generics

paskal opened the pull request:
Thread-safe. It could be used in place of simplelru.LRU but shouldn't as it has built-in locks, whereas simplelru.LRU doesn't, which allows more effective locking on top of it in top-level package cache implementations.

Similar to what was done in #41. I've tried to make the new cache as close as possible to the existing one regarding code style and behaviour, however existing simplelru.LRU is not thread-safe, as pointed out above, so it's not a drop-in replacement.

Another difference I left in place is that `size=0` on this type of cache means _unlimited_, which makes sense with the expirable cache.

Original work was done in [lcw](https://github.com/go-pkgz/lcw/tree/lru_internal), but it turned out that I don't need it there, and I thought it would be nice to add my changes to upstream (this repo).

This PR is a recreation of #68, which was not merged. cc @jefferai @Dentrax for review as an alternative to #113 (which I unlikely will be able to finish while preserving the current package's speed).

Dentrax commented on simplelru/expirable_lru.go:
```diff
@@ -0,0 +1,257 @@
+	// with non-zero TTL and size defined
+	if res.ttl != noEvictionTTL && (res.size > 0 || res.purgeEvery > 0) {
+		if res.purgeEvery <= 0 {
+			res.purgeEvery = time.Minute * 5 // non-zero purge enforced because size defined
```
Any reason why default is set to `5m`?

Dentrax commented on simplelru/expirable_lru.go:
```diff
@@ -0,0 +1,257 @@
+	return &res
+}
+
+// Purge clears the cache completely.
```
Maybe worth to mention this function also emits an event for each key.

Dentrax commented on simplelru/expirable_lru.go:
```diff
@@ -0,0 +1,257 @@
+	c.evictList.init()
+}
+
+// Add adds a value to the cache. Returns true if an eviction occurred.
```
It would be nice to mention which conditions resulting `false`.

Dentrax commented on simplelru/expirable_lru.go:
```diff
@@ -0,0 +1,257 @@
+func (c *ExpirableLRU[K, V]) Close() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	close(c.done)
```
What if chan already closed? Should we check?

Dentrax reviewed (commented):
Thanks! I think this is a better approach than #113. Do you have any new benchmark results as you already did [here](https://github.com/hashicorp/golang-lru/pull/113#issuecomment-1320586867)?

PTAL @jefferai

jefferai commented:
Is this code the same as the changed code in #113?

paskal commented:
Precisely the same code as in #113, but in a separate file, as making these changes in existing simplelru decreases the performance of the service.

Dentrax commented:
Hey @jefferai, do you have any feedback/reviews for the PR, additionally?

jefferai commented:
No, I'm just busy with personal stuff (moving) on top of work and haven't been able to get to this yet, sorry.

paskal commented:
I've completely forgot about this one, I'm addressing your comments in the meantime.

gdsoumya commented:
Is this PR getting merged anytime soon?

paskal commented:
I have to add few minor changes prior to that, I'll try to add them before the New Year.

jefferai commented:
I was hoping to look at combining the two but realistically I don't have time right now. That said there isn't really an issue with having it be separate and trying to merge them later if I see a good way. (Mostly I am just worried that fixes applied to one will be forgotten for the other.)

Please address the comments when you get a chance and I will then do another look through and hopefully get it merged.

paskal commented:
Fixed benchmarks everywhere to calculate the hit ratio correctly. Affects only logging. Benchmarks:

```
Benchmark2Q_Rand-8    	 4362295	       273.2 ns/op
Benchmark2Q_Freq-8    	 5259523	       221.2 ns/op
BenchmarkARC_Rand-8   	 3586426	       329.8 ns/op
BenchmarkARC_Freq-8   	 4349251	       263.5 ns/op
BenchmarkLRU_Rand-8   	 9076308	       133.0 ns/op
BenchmarkLRU_Freq-8   	 9300483	       128.1 ns/op
BenchmarkExpirableLRU_Rand_NoExpire-8     	 7061056	       170.3 ns/op
BenchmarkExpirableLRU_Freq_NoExpire-8     	 7251532	       165.1 ns/op
BenchmarkExpirableLRU_Rand_WithExpire-8   	 6939873	       172.0 ns/op
BenchmarkExpirableLRU_Freq_WithExpire-8   	 7980356	       143.6 ns/op
```

jefferai commented:
Curious about the choice of using a list and iterating through on a schedule vs. using a heap (e.g. adapting https://golang.google.cn/pkg/container/heap/) and having a timer keyed off of the next expiration (reset on any push if needed). That seems like it would purge more closely to timeouts.

paskal commented:
If I recall correctly, the idea is to break out of checking the list as soon as the first non-expired item is found so that the complexity of cleanup is M, where M is the number of cleaned entries. With heap by expiry date usage, the complexity of cleanup increases to M*log(N), where N is the total number of entries as we are forced to rebalance the heap each time we drop the oldest item from there.

However, I see that breaking out of the loop is not implemented there, and I have useless [continue](https://github.com/hashicorp/golang-lru/pull/116/files#diff-be4b9429f63e821595d57501426a7f91505ab8f2df186c5f5e3b0af74e1a5dfcR254): I'll fix that.

While we are speaking of the expiring mechanism, also I see that I don't remove the outdated items once I hit them on `cache.Get`, but I think it's okay as that couldn't alter the performance of existing expiry cleanup mechanism except for very special cache usage cases.

And last, in another cache implementation, I saw PurgeEvery set to half of entry TTL: It might be a better solution than the currently hardcoded 5 minutes default. What do you think?

paskal commented:
I forgot we're dealing with LRU here, and the expiry check can't stop at first non-expired elements. When items are moved to the front after access, their TTL is left intact.

So the cleanup complexity now is plain N, with the additional heap would be M*log(N), and with the additional list would be M. Currently, we have the maximal size of the cache before the clean up, but with an additional list or heap, we can trigger the cleanup call in the background once we find an expired entry so that the size would be more balanced at the cost of more frequent cleanups with locks. Basically, we're delving into the Garbage Collection area of problems:)

Another simple improvement I can add to the current setup to decrease the complexity to M is checking for `size == 0` and then not doing LRU so that the cleanup will take N for LRU expirable cache and M for the non-LRU expirable cache.

Please let me know if you want me to alter the solution and how.

jefferai commented:
So I don't have super great answers here, because anything that simplifies the mechanism necessarily trades off performance. As you noted, we're into the GC side of things.

One possibility is to actually load a timer for every single expiring entry. This has the benefit that expiration is essentially instant and you don't have to track anything, but it trades off goroutines. Vault does this, but Vault is not necessarily tracking as many objects as you might want to track here.

If we knew for sure that the key is something we could e.g. hash -- or require such a value for expiring entries -- you could use something like a heap but shard entries across some set of these. E.g. hash it, take the first byte as an index into 256 different heaps, and have a background process check each heap periodically.

Another approach is time bucketing -- keep a list or heap or something for all items expiring within some period of time. The challenge here is figuring out the bucketing size -- maybe you have 1000 entries expiring every minute, or maybe 1 entry every 1000 minutes -- but it means you only ever need to check one bucket of items when checking for expiration, and you can just run through the list.

jefferai commented:
I am worried about a goroutine per item. Nothing stops you from having a million items with a million goroutines. Even partitioning it as some libs do can leave you with a very lengthy stack trace if you SIGQUIT. go-cache for instance uses a pool of goroutines to do expiration and even then any stack trace requires filtering out goroutine after goroutine.

paskal commented:
Any other approach I can think of is more complicated to implement, read, comprehend, and troubleshoot. I know a production project with 200-250k goroutines, so the number is not a big problem.
Time buckets mean memory overhead, and with more complicated code, we're still having a big number of goroutines. For example, with buckets of size TTL/100, for 10ms it would be 100us, for 100ms - 1ms, 5m - 6s, 1h - 36s, 4h - 2m44s, 12h - 7m12s, 24h - 14m24s. What do you think?

jefferai commented:
I agree the number isn't a problem -- the troubleshooting is. A stack trace with 250k goroutines that have to be filtered out in order to get to the important bits makes any project using it much harder to troubleshoot.

Time bucketing could bring you down to a small number of goroutines -- in fact, you could just use a single one, with a timer that is reset to the expire time of the next bucket. So you could easily do 100 or 1000 buckets. I think the question then is one of philosophy for whoever wants to use this: do you care about items expiring at their expiration or not? I'm not sure if there is a clear answer, although I lean on the idea that you don't -- it's an LRU so items can be evicted at any time anyways. So probably the contract should be that you can be sure that when it's purged the callback is run, but the docs make it clear that you cannot count on it being purged at a specific time.

Thoughts?

paskal commented:
I've reverted the changes introducing per-item cleanup and created 100 buckets cleaned up every 1/100th of TTL, clarifying code and comments in the process. I've outlined the tradeoffs in the code itself: we have ~1% memory overhead and entries are cleaned up to 1% earlier than their expiration date. This cleanup interval and overhead make sense for expiring TTL on intervals varying from 100ms to 10 days.

```
$ go test -bench=.
Benchmark2Q_Rand-8       4471934               268.0 ns/op
Benchmark2Q_Freq-8       5376478               220.6 ns/op
BenchmarkARC_Rand-8      3617769               326.1 ns/op
BenchmarkARC_Freq-8      4327766               247.9 ns/op
BenchmarkLRU_Rand-8      8963338               134.0 ns/op
BenchmarkLRU_Freq-8      9173343               131.8 ns/op
BenchmarkExpirableLRU_Rand_NoExpire-8            4597234               263.2 ns/op
BenchmarkExpirableLRU_Freq_NoExpire-8            4761190               249.8 ns/op
BenchmarkExpirableLRU_Rand_WithExpire-8          4427887               277.7 ns/op
BenchmarkExpirableLRU_Freq_WithExpire-8          5001783               230.5 ns/op
```

Because `simplelru/list.go` was changed, I've checked benchmark from this repo's current master, and the numbers are the same.

I'm satisfied with the current code and the tradeoffs I've made with overhead. Please let me know what you think.

jefferai commented:
Benchmarks and such look good, but...

>entries are cleaned up to 1% earlier than their expiration date

I think it's safer to potentially clean up slightly _after_ the expiration date. In usage where you want expiration but are unlikely to drop items from the cache due to LRU behavior it might be surprising that they are gone before the expected time. Thoughts?

paskal commented:
I tried to wrap my head around that, and I couldn't. If we introduce the delay, it will increase over time, and I don't know how to keep it constant. My opinion is that a 1% early expiration is tolerable.

jefferai commented:
I'm not sure what you mean? I would imagine you're in control of the interval :-) Why will it increase over time?

jefferai commented:
@paskal Just want to follow up on the above question. Looking forward to getting this in!

paskal commented:
I still can't wrap my head around how to make it work without checking each entry expiry date which will slow everything down. I think I'll reintroduce the entry expiry date check and report back on how it affects performance.

jefferai commented:
I think we're talking about different things? You said in the current implementation you may end up cleaning up slightly before the expiration. What's so different about potentially cleaning up _after_ the expiration?

paskal commented:
@jefferai I finally found a solution which satisfies me: instead of checking each entry against `time.Now`, storing newest entry expiration date per bucket. Now we can store more expired entries (hard to predict how much precisely), but entries are cleaned up only after expiration.

Benchmark:
```
Benchmark2Q_Rand-8       4371900               278.6 ns/op
Benchmark2Q_Freq-8       5191207               225.4 ns/op
BenchmarkLRU_Rand-8      8875495               135.8 ns/op
BenchmarkLRU_Freq-8      9196723               130.7 ns/op
BenchmarkExpirableLRU_Rand_NoExpire-8            4625799               259.7 ns/op
BenchmarkExpirableLRU_Freq_NoExpire-8            4865611               247.6 ns/op
BenchmarkExpirableLRU_Rand_WithExpire-8          4312462               278.8 ns/op
BenchmarkExpirableLRU_Freq_WithExpire-8          4865191               236.8 ns/op
```

I can't test ARC anymore as it appears to be broken:
```
~/golang-lru/arc ❯ go test ./...
# github.com/hashicorp/golang-lru/arc/v2 [github.com/hashicorp/golang-lru/arc/v2.test]
./arc_test.go:24:14: undefined: getRand
./arc_test.go:53:15: undefined: getRand
./arc_test.go:55:15: undefined: getRand
./arc_test.go:84:10: undefined: getRand
./arc_test.go:85:8: undefined: getRand
FAIL    github.com/hashicorp/golang-lru/arc/v2 [build failed]
FAIL
```

jefferai commented:
Hi @paskal, ARC was moved to its own module -- it's still there, you'll just need to update against main and update your import.

paskal commented:
There is no import of ARC, I've tried just cloning the `master` and then running `cd arc; go test ./...`, and I get these errors: I can't run test or benchmarks against it. Unless I'm missing something, `getRand` used in tests should be duplicated into `arc` module.

jefferai commented:
Oop, apologies -- fixed!

paskal commented:
@jefferai, is there anything preventing us from merging this? I think I've addressed all concerns raised before, the PR is ready for the final review.

shinji62 commented:
@jefferai Would be nice to merge that, honestly that will avoid to have people to implement their own expiring lru

jefferai commented:
Nope, just need time to get back to a review. @mgaffney can look as well.

mgaffney commented on simplelru/expirable_lru.go:
```diff
@@ -0,0 +1,327 @@
+	// with non-zero TTL
+	if res.ttl != noEvictionTTL {
+		go func(done <-chan struct{}) {
+			ticker := time.NewTicker(res.ttl / numBuckets)
```
Any reason not to add a `defer ticker.Stop()` call here?

mgaffney commented on simplelru/expirable_lru.go:
```diff
@@ -0,0 +1,327 @@
+
+// Values returns a slice of the values in the cache, from oldest to newest.
+// Expired entries are filtered out.
+func (c *ExpirableLRU[K, V]) Values() []V {
```
Can we add tests for this method?

mgaffney commented:
@paskal can we move the expirable LRU code to its own package? Having `simplelru.LRU` and `expirable.LRU` types seems more idiomatic and readable to me than having `simplelru.LRU` and `simplelru.ExpirableLRU` types.

paskal commented:
@mgaffney, it is done; please take a look at the current version. I've had to duplicate the `list.go` as its unexported methods are used and should not be exposed to the users of the library.

SoMuchForSubtlety commented:
> I've had to duplicate the `list.go` as its unexported methods are used and should not be exposed to the users of the library.

It could be moved to internal, so it can be shared without being exported.

paskal commented:
This was done as well.

SoMuchForSubtlety commented on expirable/expirable_lru.go:
```diff
@@ -0,0 +1,333 @@
+	return diff
+}
+
+// Close destroys cleanup goroutine. To clean up the cache, run Purge() before Close().
+func (c *LRU[K, V]) Close() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	select {
+	case <-c.done:
+		return
+	default:
+	}
+	close(c.done)
+}
```
Having operations on a closed LRU panic would be an option. There is precedent for this behaviour in [writing to closed channels](https://go.dev/ref/spec#Close).

llorllale replied:
this would have been useful if you are caching expirable LRUs: we could have invoked `Close()` in an `EvictCallback`

mgaffney reviewed (approved):
@paskal thank you for all your hard work on this!
